dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/orgunit/handler/MetadataOrgUnitMergeHandler.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2004-2021, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.merge.orgunit.handler;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import lombok.AllArgsConstructor;
+
+import org.hisp.dhis.category.CategoryOption;
+import org.hisp.dhis.configuration.Configuration;
+import org.hisp.dhis.configuration.ConfigurationService;
+import org.hisp.dhis.dataset.DataSet;
+import org.hisp.dhis.merge.orgunit.OrgUnitMergeRequest;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.organisationunit.OrganisationUnitGroup;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserQueryParams;
+import org.hisp.dhis.user.UserService;
+import org.springframework.stereotype.Service;
+
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * Merge handler for metadata entities.
+ *
+ * @author Lars Helge Overland
+ */
+@Service
+@AllArgsConstructor
+public class MetadataOrgUnitMergeHandler
+{
+    private final UserService userService;
+
+    private final ConfigurationService configService;
+
+    public void mergeDataSets( OrgUnitMergeRequest request )
+    {
+        Set<DataSet> dataSets = request.getSources().stream()
+            .map( OrganisationUnit::getDataSets )
+            .flatMap( Collection::stream )
+            .collect( Collectors.toSet() );
```

Comment:
Wouldn't it make more sense to continue using the stream here with `forEach` below instead of collecting it to a set which is iterated using `forEach`? (Same below, maybe there is even a generic helper method that could be extracted which also would help to give the process a name)

Reply:
The reason is that the operation includes bidirectionally removing org units from these data sets (including removing data sets from the org untis). I actually did what you say first, but it leads to a `ConcurrentModificationException` since we actually change the data sets while iterating over them in the same stream.